File: app/src/main/java/de/bitb/astroskop/ui/base/BaseFragment.java
package de.bitb.astroskop.ui.base;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import javax.inject.Inject;

import butterknife.ButterKnife;
import de.bitb.astroskop.Constants;
import de.bitb.astroskop.R;
import de.bitb.astroskop.enums.AnimationType;
import de.bitb.astroskop.injection.IBind;
import de.bitb.astroskop.injection.IInjection;
import de.bitb.astroskop.viewbuilder.DialogBuilder;
import de.bitb.astroskop.viewbuilder.ToastBuilder;

import static de.bitb.astroskop.Constants.NULL_INTEGER;

public abstract class BaseFragment extends Fragment implements IBaseView {

    @Inject
    protected ToastBuilder toastUtils;

    @Inject
    protected DialogBuilder dialogBuilder;

    private IToolbarView toolbarView;
    private ActionbarHandler actionbarHandler;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        this.actionbarHandler = new ActionbarHandler(getActionbarCallback());
        if (this instanceof IInjection) {
            ((IInjection) this).inject(((BaseActivity) getActivity()).getAppComponent());
        } else {
            ((BaseActivity) getActivity()).getAppComponent().inject(this);
        }
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);

        View root = inflater.inflate(getLayoutId(), null, false);
        if (this instanceof IBind) {
            ButterKnife.bind(this, root);
        }

        setHasOptionsMenu(true);
        return root;
    }

    public abstract int getLayoutId();

    public ActionbarHandler.ActionbarCallback getActionbarCallback() {
        return new ActionbarHandler.ActionbarCallback();
    }

    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        actionbarHandler.onPrepareOptionsMenu(menu);
        super.onPrepareOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        return actionbarHandler.onOptionsItemSelected(item.getItemId()) || super.onOptionsItemSelected(item);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof IToolbarView) {
            toolbarView = (IToolbarView) context;
        }
    }

    protected void runOnUiThread(Runnable runnable) {
        if (getActivity() != null) {
            getActivity().runOnUiThread(runnable);
        }
    }

    public void setToolbarTitle(String title) {
        if (toolbarView != null) {
            toolbarView.setToolbarTitle(title);
        }
    }

    public boolean allowBackPressed() {
        return true;
    }

    public AnimationType getAnimationType() {
        return AnimationType.NONE;
    }

}

File: app/src/main/java/de/bitb/astroskop/ui/base/BaseFragment.kt
package de.bitb.astroskop.ui.base

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.*
import butterknife.ButterKnife
import de.bitb.astroskop.enums.AnimationType
import de.bitb.astroskop.injection.IBind
import de.bitb.astroskop.injection.IInjection
import de.bitb.astroskop.viewbuilder.DialogBuilder
import de.bitb.astroskop.viewbuilder.ToastBuilder
import javax.inject.Inject


abstract class BaseFragment : Fragment(), IBaseView {

    @Inject
    protected var toastUtils: ToastBuilder? = null

    @Inject
    protected var dialogBuilder: DialogBuilder? = null

    private var toolbarView: IToolbarView? = null
    var actionbarHandler: ActionbarHandler? = null

    abstract val layoutId: Int

    var allowBackPress: Boolean = false

    protected open val actionbarCallback: ActionbarHandler.ActionbarCallback
        get() = ActionbarHandler.ActionbarCallback()

    val animationType: AnimationType
        get() = AnimationType.NONE

    override fun onCreate(savedInstanceState: Bundle?) {
        this.actionbarHandler = ActionbarHandler(actionbarCallback)
        if (this is IInjection) {
            (this as IInjection).inject((activity as BaseActivity).appComponent)
        } else {
            (activity as BaseActivity).appComponent!!.inject(this)
        }
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)

        val root = inflater.inflate(layoutId, null, false)
        if (this is IBind) {
            ButterKnife.bind(this, root)
        }

        setHasOptionsMenu(true)
        return root
    }

    override fun onPrepareOptionsMenu(menu: Menu?) {
        actionbarHandler!!.onPrepareOptionsMenu(menu)
        super.onPrepareOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        return actionbarHandler!!.onOptionsItemSelected(item!!.itemId) || super.onOptionsItemSelected(item)
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is IToolbarView) {
            toolbarView = context
        }
    }

    protected fun runOnUiThread(runnable: Runnable) {
        activity?.runOnUiThread(runnable)
    }

    fun setToolbarTitle(title: String) {
        toolbarView?.setToolbarTitle(title)
    }

}
